Unable to resume cache when running on AWS Batch when there are spaces in file names

When running jobs on AWS Batch via Nextflow tower and there are spaces in the file names, the workflow finds the files and runs successfully. However, when I try to resume the workflow, it is unable to resume and starts the workflow from beginning. Checking the log files reveals an illegal character error: 

```
Mar-17 08:59:16.521 [Actor Thread 6] WARN  nextflow.processor.TaskProcessor - [FG_GATK_BAM:PROCESS_BAM:PROCESS_MARKDUPS_ONLY:BWA_MEM_PAIRED (test_sample-1_1)] Unable to resume cached task -- See log file for details
com.esotericsoftware.kryo.KryoException: java.lang.IllegalArgumentException: Illegal character in path at index 54: s3:///kc1-test-output/testing_filename_spaces/SRR6798772 toy10k 1.fastq.gz
Serialization trace:
target (nextflow.processor.TaskPath)
	at com.esotericsoftware.kryo.serializers.ObjectField.read(ObjectField.java:125)
	at com.esotericsoftware.kryo.serializers.FieldSerializer.read(FieldSerializer.java:528)
	at com.esotericsoftware.kryo.Kryo.readClassAndObject(Kryo.java:761)
	at com.esotericsoftware.kryo.serializers.MapSerializer.read(MapSerializer.java:143)
	at com.esotericsoftware.kryo.serializers.MapSerializer.read(MapSerializer.java:21)
	at com.esotericsoftware.kryo.Kryo.readClassAndObject(Kryo.java:761)
	at com.esotericsoftware.kryo.Kryo$readClassAndObject$16.call(Unknown Source)
	at nextflow.util.KryoHelper.deserialize(SerializationHelper.groovy:182)
	at nextflow.util.KryoHelper.deserialize(SerializationHelper.groovy)
	at nextflow.util.KryoHelper$deserialize$4.call(Unknown Source)
	at nextflow.processor.TaskContext.deserialize(TaskContext.groovy:190)
	at nextflow.CacheDB.getTaskEntry(CacheDB.groovy:163)
	at nextflow.CacheDB$getTaskEntry.call(Unknown Source)
	at nextflow.processor.TaskProcessor.checkCachedOutput(TaskProcessor.groovy:869)
	at nextflow.processor.TaskProcessor.checkCachedOrLaunchTask(TaskProcessor.groovy:758)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.codehaus.groovy.runtime.callsite.PlainObjectMetaMethodSite.doInvoke(PlainObjectMetaMethodSite.java:43)
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite$PogoCachedMethodSiteNoUnwrapNoCoerce.invoke(PogoMetaMethodSite.java:193)
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite.callCurrent(PogoMetaMethodSite.java:61)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:203)
	at nextflow.processor.TaskProcessor.invokeTask(TaskProcessor.groovy:592)
	at nextflow.processor.InvokeTaskAdapter.call(InvokeTaskAdapter.groovy:59)
	at groovyx.gpars.dataflow.operator.DataflowOperatorActor.startTask(DataflowOperatorActor.java:120)
	at groovyx.gpars.dataflow.operator.ForkingDataflowOperatorActor.access$001(ForkingDataflowOperatorActor.java:35)
	at groovyx.gpars.dataflow.operator.ForkingDataflowOperatorActor$1.run(ForkingDataflowOperatorActor.java:58)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.IllegalArgumentException: Illegal character in path at index 54: s3:///kc1-test-output/testing_filename_spaces/SRR6798772 toy10k 1.fastq.gz
	at java.base/java.net.URI.create(URI.java:883)
	at nextflow.util.PathSerializer.read(SerializationHelper.groovy:284)
	at nextflow.util.PathSerializer.read(SerializationHelper.groovy)
	at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:679)
	at com.esotericsoftware.kryo.serializers.ObjectField.read(ObjectField.java:106)
	... 30 common frames omitted
```
Ideally, there wouldn't be any spaces in the file names, however we receive files from external sources with spaces in and do not want to rename them. Additionally, this issue doesn't occur on our local servers where spaces do not pose an issue when resuming. Is there a way to escape spaces so that Nextflow both recognises the files when running the first time AND when resuming a run on AWS Batch?

--- a/modules/nextflow/src/main/groovy/nextflow/util/SerializationHelper.groovy
+++ b/modules/nextflow/src/main/groovy/nextflow/util/SerializationHelper.groovy
@@ -280,8 +280,12 @@ class PathSerializer extends Serializer<Path> {
             return FileSystems.getDefault().getPath(path)
         }
 
-        // try to find provider
-        def uri = URI.create("$scheme://${path.replace(' ', '%20')}")
+        // NOTE: PATHS FOR NON DEFAULT FILE SYSTEMS SHOULD BE CREATED
+        // BY THE CORRESPONDING SPECIALISED SERIALIZER CLASSES
+        //
+        // THIS IS LEGACY CODE THAT EVENTUALLY SHOULD BE REMOVED
+        //
+        def uri = URI.create("$scheme://$path")
         def fs = FileHelper.getOrCreateFileSystemFor(uri)
         return fs.provider().getPath(uri)
     }

--- a/modules/nextflow/src/main/groovy/nextflow/util/SerializationHelper.groovy
+++ b/modules/nextflow/src/main/groovy/nextflow/util/SerializationHelper.groovy
@@ -281,7 +281,7 @@ class PathSerializer extends Serializer<Path> {
         }
 
         // try to find provider
-        def uri = URI.create("$scheme://$path")
+        def uri = URI.create("$scheme://${path.replace(' ', '%20')}")
         def fs = FileHelper.getOrCreateFileSystemFor(uri)
         return fs.provider().getPath(uri)
     }

--- a/plugins/nf-amazon/src/main/nextflow/cloud/aws/util/S3PathFactory.groovy
+++ b/plugins/nf-amazon/src/main/nextflow/cloud/aws/util/S3PathFactory.groovy
@@ -18,9 +18,7 @@ class S3PathFactory extends FileSystemPathFactory {
         // normalise 's3' path
         if( str.startsWith('s3://') && str[5]!='/' ) {
             final path = "s3:///${str.substring(5)}"
-            // note: this URI constructor parse the path parameter and extract the `scheme` and `authority` components
-            final uri = new URI(null,null, path,null,null)
-            return FileHelper.getOrCreateFileSystemFor(uri).provider().getPath(uri)
+            return create(path)
         }
         return null
     }
@@ -42,4 +40,22 @@ class S3PathFactory extends FileSystemPathFactory {
                 : null
     }
 
+    /**
+     * Creates a {@link S3Path} from a S3 formatted URI.
+     *
+     * @param path
+     *      A S3 URI path e.g. s3:///BUCKET_NAME/some/data.
+     *      NOTE it expect the s3 prefix provided with triple `/` .
+     *      This is required by the underlying implementation expecting the host name in the URI to be empty
+     *      and the bucket name to be the first path element
+     * @return
+     *      The corresponding {@link S3Path}
+     */
+    static S3Path create(String path) {
+        if( !path ) throw new IllegalArgumentException("Missing S3 path argument")
+        if( !path.startsWith('s3:///') ) throw new IllegalArgumentException("S3 path must start with s3:/// prefix -- offending value '$path'")
+        // note: this URI constructor parse the path parameter and extract the `scheme` and `authority` components
+        final uri = new URI(null,null, path,null,null)
+        return (S3Path)FileHelper.getOrCreateFileSystemFor(uri).provider().getPath(uri)
+    }
 }

--- a/plugins/nf-amazon/src/main/nextflow/cloud/aws/util/S3PathSerializer.groovy
+++ b/plugins/nf-amazon/src/main/nextflow/cloud/aws/util/S3PathSerializer.groovy
@@ -17,22 +17,47 @@
 
 package nextflow.cloud.aws.util
 
+
+import com.esotericsoftware.kryo.Kryo
+import com.esotericsoftware.kryo.Serializer
+import com.esotericsoftware.kryo.io.Input
+import com.esotericsoftware.kryo.io.Output
 import com.upplication.s3fs.S3Path
 import groovy.transform.CompileStatic
-import nextflow.util.PathSerializer
+import groovy.util.logging.Slf4j
 import nextflow.util.SerializerRegistrant
 import org.pf4j.Extension
-
 /**
  * Register the S3Path serializer
  *
  * @author Paolo Di Tommaso <paolo.ditommaso@gmail.com>
  */
+@Slf4j
 @Extension
 @CompileStatic
-class S3PathSerializer implements SerializerRegistrant  {
+class S3PathSerializer extends Serializer<S3Path> implements SerializerRegistrant  {
+
     @Override
     void register(Map<Class, Object> serializers) {
-        serializers.put(S3Path, PathSerializer)
+        serializers.put(S3Path, S3PathSerializer)
     }
+
+    @Override
+    void write(Kryo kryo, Output output, S3Path target) {
+        final scheme = target.getFileSystem().provider().getScheme()
+        final path = target.toString()
+        log.trace "S3Path serialization > scheme: $scheme; path: $path"
+        output.writeString(scheme)
+        output.writeString(path)
+    }
+
+    @Override
+    S3Path read(Kryo kryo, Input input, Class<S3Path> type) {
+        final scheme = input.readString()
+        final path = input.readString()
+        if( scheme != 's3' ) throw new IllegalStateException("Unexpected scheme for S3 path -- offending value '$scheme'")
+        log.trace "S3Path de-serialization > scheme: $scheme; path: $path"
+        return (S3Path) S3PathFactory.create("s3://${path}")
+    }
+    
 }

--- a/plugins/nf-amazon/src/test/nextflow/util/S3PathSerializerTest.groovy
+++ b/plugins/nf-amazon/src/test/nextflow/util/S3PathSerializerTest.groovy
@@ -1,0 +1,46 @@
+/*
+ * Copyright 2020-2022, Seqera Labs
+ * Copyright 2013-2019, Centre for Genomic Regulation (CRG)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package nextflow.util
+
+import nextflow.cloud.aws.util.S3PathFactory
+import spock.lang.Specification
+/**
+ *
+ * @author Ben Sherman <bentshermann@gmail.com>
+ */
+class S3PathSerializerTest extends Specification {
+
+    def 'should serialise s3 path' () {
+        when:
+        def path = S3PathFactory.parse('s3://mybucket/file.txt')
+        def buffer = KryoHelper.serialize(path)
+        then:
+        KryoHelper.deserialize(buffer).getClass().getName() == 'com.upplication.s3fs.S3Path'
+        KryoHelper.deserialize(buffer) == S3PathFactory.parse('s3://mybucket/file.txt')
+    }
+
+    def 'should serialise s3 path with spaces' () {
+        when:
+        def path = S3PathFactory.parse('s3://mybucket/file with spaces.txt')
+        def buffer = KryoHelper.serialize(path)
+        then:
+        KryoHelper.deserialize(buffer).getClass().getName() == 'com.upplication.s3fs.S3Path'
+        KryoHelper.deserialize(buffer) == S3PathFactory.parse('s3://mybucket/file with spaces.txt')
+    }
+
+}

